pkg/crds/crds.go

```diff
@@ -0,0 +1,193 @@
+		items, err := clientsets.ListResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace)
+		if err != nil {
+			if k8sErrors.IsNotFound(err) {
+				logging.Info("the server could not find the requested resource: %s", resource)
+				continue
+			}
+			return err
+		}
+
+		if len(items) == 0 {
+			logging.Info("resource %s was not found on the cluster", resource)
+			continue
+		}
+
+		for _, item := range items {
+			logging.Info(fmt.Sprintf("removing resource %s: %s", resource, item.GetName()))
+			err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+			if err != nil {
+				if k8sErrors.IsNotFound(err) {
+					logging.Info(err.Error())
+					continue
+				}
+				return err
+			}
+
+			itemResource, err := clientsets.GetResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, item.GetName(), clusterNamespace)
+			if err != nil {
+				if !k8sErrors.IsNotFound(err) {
+					return err
+				}
+			}
+
+			if itemResource != nil {
+				logging.Info(fmt.Sprintf("resource %q is not yet deleted, applying patch to remove finalizer...", itemResource.GetName()))
+				err = updatingFinalizers(ctx, clientsets, itemResource, resource, clusterNamespace)
+				if err != nil {
+					if k8sErrors.IsNotFound(err) {
+						logging.Info(err.Error())
+						continue
+					}
+					return err
+				}
+
+				err = clientsets.DeleteResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, item.GetName())
+				if err != nil {
+					if !k8sErrors.IsNotFound(err) {
+						return err
+					}
+				}
+			}
+			logging.Info(fmt.Sprintf("resource %s was deleted", item.GetName()))
+		}
+	}
+	return nil
+}
+
+func updatingFinalizers(ctx context.Context, clientsets k8sutil.ClientsetsInterface, itemResource *unstructured.Unstructured, resource, clusterNamespace string) error {
+	patch := defaultResourceRemoveFinalizers
+
+	if itemResource.GetKind() == CephResourceKindCephCluster {
+		patch = clusterResourcePatchFinalizer
+	}
+
+	jsonPatchData, _ := json.Marshal(patch)
+
+	err := clientsets.PatchResourcesDynamically(ctx, cephRookIoGroup, cephRookResourcesVersion, resource, clusterNamespace, itemResource.GetName(), types.MergePatchType, jsonPatchData)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func ensureClusterIsEmpty(ctx context.Context, k8sClientSet kubernetes.Interface, clusterNamespace string) {
+	timer := time.NewTimer(timeOutCleanupResources)
+	for {
+		deploymentsFound := 0
+		deployments, err := k8sClientSet.AppsV1().Deployments(clusterNamespace).List(ctx, v1.ListOptions{})
+		if err != nil {
+			logging.Fatal(err)
+		}
+		for _, deployment := range deployments.Items {
+			if strings.EqualFold(deployment.Name, rookOperatorDeployment) {
+				continue
+			}
+			deploymentsFound++
+		}
+
+		if deploymentsFound == 0 {
+			break
+		}
+		logging.Info("still working on cleaning up resources")
```

Comment:
Could we show which deployments we are waiting on? Perhaps a comma-separated list?
```suggestion
		logging.Info("waiting to clean up resources: %s", deploymentNames)
```